server/teamshandler_test.go

```diff
@@ -0,0 +1,349 @@
+		requestData := v1structs.UpsertTeamRequest{
+			TeamRoster:               goodRoster,
+			ArmoredDetachedSignature: goodSignature,
+		}
+
+		response := callAPIWithJSON(t, "POST", "/v1/teams", requestData, &signerFingerprint)
+
+		t.Run("returns HTTP 201", func(t *testing.T) {
+			assertStatusCode(t, http.StatusCreated, response.Code)
+		})
+
+		t.Run("adds valid database row", func(t *testing.T) {
+			team, err := datastore.GetTeam(nil, goodUUID)
+			assert.ErrorIsNil(t, err)
+
+			assert.Equal(t, goodUUID, team.UUID)
+			assert.Equal(t, goodRoster, team.Roster)
+			assert.Equal(t, goodSignature, team.RosterSignature)
+
+			if now := time.Now(); team.CreatedAt.Sub(now) > time.Duration(5)*time.Second {
+				t.Fatalf("expected team.CreatedAt to be within 5s of now (%s), got %s",
+					now, team.CreatedAt)
+			}
+		})
+	})
+
+	t.Run("request doesn't contain signer fingerprint in auth header", func(t *testing.T) {
+		requestData := v1structs.UpsertTeamRequest{
+			TeamRoster:               goodRoster,
+			ArmoredDetachedSignature: goodSignature,
+		}
+
+		response := callAPIWithJSON(
+			t, "POST", "/v1/teams", requestData, nil) // nil -> unauthenticated
+		assertStatusCode(t, http.StatusBadRequest, response.Code)
+		assertHasJSONErrorDetail(t,
+			response.Body,
+			"missing Authorization header starting `tmpfingerprint: OPENPGP4FPR:`")
+	})
+
+	testEndpointRejectsBadJSON(t, "POST", "/v1/teams")
+
+	t.Run("missing json field", func(t *testing.T) {
+		t.Run("teamRoster", func(t *testing.T) {
+			requestData := v1structs.UpsertTeamRequest{
+				ArmoredDetachedSignature: goodSignature,
+			}
+
+			response := callAPIWithJSON(t, "POST", "/v1/teams", requestData, &signerFingerprint)
+			assertStatusCode(t, http.StatusBadRequest, response.Code)
+			assertHasJSONErrorDetail(t, response.Body, "missing teamRoster")
+		})
+
+		t.Run("armoredDetachedSignature", func(t *testing.T) {
+			requestData := v1structs.UpsertTeamRequest{
+				TeamRoster: goodRoster,
+			}
+
+			response := callAPIWithJSON(t, "POST", "/v1/teams", requestData, &signerFingerprint)
+			assertStatusCode(t, http.StatusBadRequest, response.Code)
+			assertHasJSONErrorDetail(t, response.Body, "missing armoredDetachedSignature")
+		})
+	})
+
+	t.Run("request doesn't contain signer fingerprint in auth header", func(t *testing.T) {
+		requestData := v1structs.UpsertTeamRequest{
+			TeamRoster:               goodRoster,
+			ArmoredDetachedSignature: goodSignature,
+		}
+
+		response := callAPIWithJSON(
+			t, "POST", "/v1/teams", requestData, nil) // nil -> unauthenticated
+		assertStatusCode(t, http.StatusBadRequest, response.Code)
+		assertHasJSONErrorDetail(t,
+			response.Body,
+			"missing Authorization header starting `tmpfingerprint: OPENPGP4FPR:`")
+	})
+
+	t.Run("mismatch between signer fingerprint and long keyID in signature", func(t *testing.T) {
+		mismatchedFingerprint := exampledata.ExampleFingerprint2 // 2 is in the API
```

Comment:
I get how putting all the setup of the API upfront in `setup` is useful, but it does make reading particular cases _marginally_ tricker. Especially since `ExampleFingerprint2` is only used in this one test case.

Reply:
Absolutely true for this example of `ExampleFingerprint2`: I could move that into this test case specifically.

But for the rest of the stuff in setup, which applies to most or all tests, I don't see that it could live anywhere else?

Reply:
No, agreed 💯